### ISSUE TITLE
docker compose does not create volume folders when the folder does not exist

### ISSUE THREAD

**Issue by dellagustin-sap** [docker/compose#8533]
**Description**

`docker compose up` for a service that describe a volume bound to a folder that does not exist yields the following error:

```
❯ docker compose up -d
[+] Running 0/0
 ⠋ Container docker-compose-non-existent-folder-binding-issue_hello-world_1  Recreate                                       0.0s
Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /<path omitted>/logs
```

`docker-compose up` will in turn create the folder used for binding in case it does not exist.

**Steps to reproduce the issue:**
1. Create a compose file with the following content:
    ```yml
    version: '2'
    services:
      hello-world:
        image: hello-world
        volumes:
        - './logs:/tmp/logs'
    ```
2. Ensure that the `logs` folder does not exist inside the folder where you created the compose file
3. Run the command `docker compose up`

**Describe the results you received:**
Already mentioned in the description

**Describe the results you expected:**
The logs folder should be created in case it does not exist

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Cloud integration: 1.0.14
 Version:           20.10.6
 API version:       1.41
 Go version:        go1.16.3
 Git commit:        370c289
 Built:             Fri Apr  9 22:46:57 2021
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.6
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8728dd2
  Built:            Fri Apr  9 22:44:56 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker context show`:**  
You can also run `docker context inspect context-name` to give us more details but don't forget to remove sensitive content.

```
default 
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  compose: Docker Compose (Docker Inc., 2.0.0-beta.1)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 9
  Running: 6
  Paused: 0
  Stopped: 3
 Images: 179
 Server Version: 20.10.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc version: 12644e614e25b05da6fd08a38ffa0cfe1903fdec
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.25-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 15.64GiB
 Name: docker-desktop
 ID: OU4T:KJXL:GFUU:COJX:NPDY:SBVL:TPCQ:JBRP:WK5F:Q2XX:CZQI:SXSF
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 84
  Goroutines: 79
  System Time: 2021-06-23T09:34:20.161579016Z
  EventsListeners: 4
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS ECS, Azure ACI, local, etc.):**

**Comment by Software-Noob** [comment]
I encountered the same issue with volumes that have trailing slashes. Removing trailing slashes from paths fixes the volumes, or using compose V1 to create the volumes, then it will work with compose 2.

This is on arch Linux. Tried 2.0.0.1 and also the Aur built package of 2.1.1-1

**Comment by mibuthu** [comment]
Same problem here.
Removing the trailing slashes from the paths as described by @Software-Noob fixes the issue.
In my opinion this is a bug, because it should also work with trailing slashes in the paths.

**Comment by Paraphraser** [comment]
Cross-linking to [PiHole issue 947](https://github.com/pi-hole/docker-pi-hole/issues/947) which is the same problem.

**Comment by dellagustin-sap** [comment]
> You're using docker compose 2.0.0-beta.1, can you please upgrade to recently released 2.0.0-beta.4 to confirm this issue is still relevant? (just need to download relevant binary for your platform from https://github.com/docker/compose-cli/releases/tag/v2.0.0-beta.4, rename as `~/.docker/docker-compose` and make executable with `chmod +x`)

Hello @ndeloof , sorry for my very late reply, I retested today having Docker Compose v2.2.1 installed on my machine.
It worked for both `- './logs:/tmp/logs'` and `- './logs/:/tmp/logs/'`, from my side as reporter, I consider this fixed.

@Software-Noob , @mibuthu , @Paraphraser , still an issue for you?

**Comment by david-a** [comment]
For those who stumble upon this issue - just add `bind.create_host_path` to your volumes section -

```
volumes:
      - type: bind
        source: localFolder/subFolderIfNeeded
        target: /data
        bind:
          create_host_path: true
```

**Comment by michealroberts** [comment]
@david-a That does not work at all. Please can you update you comment with an explanation as to when and where this works.

**Comment by Davut-Atajanov** [comment]
The issue still persists on Debian 12 (what I use)

**Comment by Paraphraser** [comment]
I'm not sure I'd call it an "issue" which "still persists".

Let me explain. My post in Nov 2021 had to do with trailing slashes on the external path of a bind-mount volume mapping causing the mount to fail. The workaround was to remove the trailing slash. That bug in `docker compose` was fixed.

Now we come to today and how you declare bind-mounts. Functionally, these two are identical:

* one-line syntax:

	```
	services:
	
	  some-service:
	    ...
	    volumes:
	      - ./volumes/vol1:/internal/vol1
	```

* multi-line syntax:

	```
	services:
	
	  some-service:
	    ...
	    volumes:
	      - type: bind
	        source: ./volumes/vol1
	        target: /internal/vol1
	        bind:
	          create_host_path: true
	```

If you omit the last two lines from the multi-line syntax then `docker compose` does not automatically create the external path. This is documented [here](https://docs.docker.com/reference/compose-file/services/#volumes), which is where I took this screen-shot:

![Image](https://github.com/user-attachments/assets/711baf0b-4cef-439f-b9fc-83829901ed18)

<hr>

# Tests

As far as I can see, `docker compose` behaves according to its documentation.

## scenario

Starting position:

```
$ grep PRETTY_NAME /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"

$ tree
.
└── docker-compose.yml

1 directory, 1 file

$ cat docker-compose.yml 
---

services:

  ubuntu:
    container_name: ubuntu
    image: ubuntu
    command: sleep 3600
    volumes:
#     - ./volumes/vol1:/internal/vol1
      - type: bind
        source: ./volumes/vol1
        target: /internal/vol1
        x-bind:
          create_host_path: true

```

In words:

* a one-line-syntax bind-mount is commented-out
* a multi-line-syntax bind mount is active; BUT
* the `x-` prefix on the second-to-last line has the effect of commenting-out those last two lines.

## Test 1

Try to start the container:

```
$ docker compose up -d
[+] Running 1/1
 ✔ Network volume-test_default  Created                                                                                                0.1s 
 ⠋ Container ubuntu             Creating                                                                                               0.0s 
Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /home/moi/volume-test/volumes/vol1
```

Show that there were no side-effects:

```
$ tree
.
└── docker-compose.yml

1 directory, 1 file
```

Conclusion:

* in the absence of the `create_host_path` statement, multi-line syntax does **not** create the external path so the bind-mount fails.

## Test 2

Activate the `create_host_path` statement and try again:

```
$ sed -i.bak s/x-bind/bind/ docker-compose.yml 

$ cat docker-compose.yml
---

services:

  ubuntu:
    container_name: ubuntu
    image: ubuntu
    command: sleep 3600
    volumes:
#     - ./volumes/vol1:/internal/vol1
      - type: bind
        source: ./volumes/vol1
        target: /internal/vol1
        bind:
          create_host_path: true

$ docker compose up -d
[+] Running 1/1
 ✔ Container ubuntu  Started                                                                                                           0.3s 

$ tree
.
├── docker-compose.yml
├── docker-compose.yml.bak
└── volumes
    └── vol1

3 directories, 2 files
```

Conclusion:

* in the presence of the `create_host_path` statement, multi-line syntax **does** create the external path so the bind-mount succeeds

## Test 3

Reset:

```
$ docker compose down
[+] Running 2/2
 ✔ Container ubuntu             Removed                                                                                               10.2s 
 ✔ Network volume-test_default  Removed                                                                                                0.3s 

$ sudo rm -rf volumes/

$ tree
.
├── docker-compose.yml
└── docker-compose.yml.bak

1 directory, 2 files
```

Activate the one-line form of bind mount and remove the multi-line form:

```
$ sed -i.bak -e 's/#/ /' -e '/- type: bind/Q' docker-compose.yml

$ cat docker-compose.yml
---

services:

  ubuntu:
    container_name: ubuntu
    image: ubuntu
    command: sleep 3600
    volumes:
      - ./volumes/vol1:/internal/vol1

$ docker compose up -d
[+] Running 2/2
 ✔ Network volume-test_default  Created                                                                                                0.1s 
 ✔ Container ubuntu             Started                                                                                                0.3s 

$ tree
.
├── docker-compose.yml
├── docker-compose.yml.bak
└── volumes
    └── vol1

3 directories, 2 files
```

Conclusion:

* one-line syntax **does** create the external path so the bind-mount succeeds.

## Bottom line

Unless you have a good **reason** for using multi-line syntax, just go with one-line syntax.

**Comment by ViktorTrojan** [comment]
Im having this issue on ubuntu server 22.04, I dont think this is really a docker / docker compose issue, I think somehow my docker group or something does not have the correct permissions for the containers to create volumes. Even when I create a volume bind to a folder, the content of the bind mount is not reflected on my host machine, and I have this issue with every container.

Can someone please tell me what the correct permissions for docker have to be or how to fix my issue?

**Comment by Paraphraser** [comment]
I honestly have no idea what "ubuntu server 22.04" is or how it differs from one of my Proxmox-VE guests which reports "Ubuntu 24.04.1 LTS" so please view what I say through that lens.

I spend a lot of time helping folks on the [IOTstack](https://github.com/SensorsIot/IOTstack) repo and its associated Discord channel. A fair number of the problems turn out to be caused by improper installation of docker and docker-compose.

I have no idea *how* you installed docker and docker-compose so I really can't say whether an "improper" installation of docker might explain the behaviour you are seeing. It may well be that your docker installation is 100% correct and that it's something about "ubuntu server" which is causing trouble. All I can really do is tell you how docker should be installed and point you to some resources - with the caveat that most of what I say is based on Debian not Ubuntu.

As a quick test, try running:

```
$ docker version -f "{{.Server.Version}}"
$ docker compose version
$ docker-compose version
$ groups
```

The answers you should get (as of the date of this post) are:

* for docker: 27.3.1
* for both methods of docker compose: v2.29.7
* the group `docker` should be in the list

If the only problem is that `docker` isn't in the groups list, try adding it:

```
$ usermod -G docker -a $USER
$ exit
```

> Side-note: one of the **consequential** problems of the current user not being in the `docker` group is a tendency, at least among novice IOTstack users, to employ `sudo` to overcome permission issues. Any time you run `sudo docker` or `sudo docker-compose` you create an absolute mess which you then have to unpick. On IOTstack, I usually advise a clean install.

Now I'll come to the various version problems. The absolute worst case scenario (at least on Debian) is if `docker` reports a version that looks like `20.10.5+dfsg1`. What that means is you actually have such an obsolete version of `docker-compose` that it has forcibly downgraded `docker` and pinned it to an ancient version. The only solution is to nuke all the Docker stuff and start over. I'll come back to this.

> as background, docker-compose has been through at least three iterations, a compiled binary (version numbers around 1.25), then a Python script (version numbers around 1.29), and now a compiled binary plugin (version numbers 2 and up).

Another common pattern is `docker` is up-to-date but any of the following is true:

1. `docker-compose version` works but `docker compose version` doesn't
2. `docker compose version` works but `docker-compose version` doesn't
3. both forms of the command work but produce different answers

In general (1) usually means you haven't installed docker using the official "convenience script" (because that installs the plugin). (2) means you probably used the convenience script but didn't do the manual step of creating a symlink so `docker-compose` points to the plugin. (3) usually implies an older install of `docker-compose` (the command) plus either a manual install of the plugin or running the convenience script on top of an existing installation.

Are you starting to see spaghetti everywhere?

If you run the commands I gave you above and the answers you get suggest you have one of these problems then there are a couple of approaches you can use to fix everything.

## fix option 1

The [PiBuilder](https://github.com/Paraphraser/PiBuilder/tree/master) repo and, in particular, the [helpers](https://github.com/Paraphraser/PiBuilder/tree/master/boot/scripts/helpers) folder has two scripts that could be useful:

* `uninstall_docker.sh`; and
* `uninstall_docker-compose.sh`

which you should run in that order and then reboot.
	
The PiBuilder [scripts](https://github.com/Paraphraser/PiBuilder/tree/master/boot/scripts) folder has:

* `04_setup.sh` which you should probably study, rather than run because it's actually intended to be run in sequence with the other scripts in that folder **and** those scripts expect a clean installation of Debian. The core of it boils down to running the convenience script to install both `docker` and `docker-compose-plugin`:

	```
	$ curl -fsSL https://get.docker.com | sudo sh
	```
	
	Create a symlink so the `docker-compose` command works:
	
	```
	$ COMPOSE_PLUGIN="/usr/libexec/docker/cli-plugins/docker-compose"
	$ COMPOSE_SYMLINK="/usr/local/bin/docker-compose"
	$ sudo ln -s "$COMPOSE_PLUGIN" "$COMPOSE_SYMLINK"
	```

	Add the current user to the Docker group:
	
	```
	$ sudo /usr/sbin/usermod -G docker -a $USER
	```

	Then finish off with a reboot.

## fix option 2

The [IOTstack](https://github.com/SensorsIot/IOTstack) repo has an `install.sh` script which is what I call a "head-banger" (if you bash your head against a brick wall long enough, you'll eventually punch a hole in the wall).

The target environment of the script is the same as PiBuilder (getting IOTstack working on Debian) but `install.sh` is intended to work with both green-fields and existing installations. The only real side-effect is that it will clone the IOTstack repo if it doesn't exist already.

Unlike the two `uninstall_` scripts above, the `install.sh` doesn't ever uninstall anything. It just tells you what you need to do.

Once everything is to the script's liking, it will install docker, the plugin, the symlink and the group.

Basically, you just keep re-running the script and following its instructions until it reports "install.sh completed".

The goal of both methods is the same. Pick whichever you like.

## A bit of proof

Some commands (with most output suppressed so I can get everything into a screen-shot):

```
$ grep "^PRETTY" /etc/os-release
$ ls
$ curl -fsSL https://raw.githubusercontent.com/SensorsIot/IOTstack/master/install.sh | bash 2>/dev/null | grep "install.sh completed"
$ docker version -f "{{.Server.Version}}"
$ docker compose version
$ docker-compose version
$ groups
$ git clone https://github.com/Paraphraser/docker-volume-explorer.git >/dev/null 2>&1
$ ls
$ cd docker-volume-explorer
$ ./explore_bind_mounts.sh 
$ ./run_test_with_self_repair.sh false >/dev/null 2>&1
$ ./explore_bind_mounts.sh 
```

![Image](https://github.com/user-attachments/assets/6460cdf7-a72f-4231-9fb8-00a6de641ea2)

In words:

1. Prove the distro.
2. Show `~` is empty.
3. Run the IOTstack install script. I'm capturing the `install.sh completed` because that's the signature of the script saying Docker is installed properly.
4. Show the versions of docker, the plugin, and the symlink to the plugin (ie same thing).
5. Show the current user is a member of the `docker` group.
6. Clone a test repo.
7. Show what's in home (the test repo and IOTstack as a side-effect of its install script).
8. Move into the test folder.
9. Show no bind-mounts exist.
10. Run the test.
11. Show bind-mounts exist.

Now, if you take yourself to the [docker-volume-explorer](https://github.com/Paraphraser/docker-volume-explorer) repo and have a look at the `docker-compose.yml` you'll see that volumes 2, 5 and 8 are one-liner bind-mounts.

So that's "Ubuntu 24.04.1 LTS". Whether it also works on "ubuntu server 22.04" is something I don't know.

**Comment by dferenc-netwizzy** [comment]
Is there a difference between "service-level" and "top-level" volume definitions?
The top-level [documentation](https://docs.docker.com/reference/compose-file/volumes/) is really short, and it seems that using the same declaration on windows the missing folders are created, while on linux they are not.

```yaml
name: testing_volumes
volumes:
  volume-mount:
    driver: local
    driver_opts:
      type: none
      # For windows "C:\mnt\test"
      device: /mnt/test
      o: bind
```

Both `C:\mnts` and `/mnt` exists with permission `777`, but while on windows (`Docker version 29.0.1, build eedd969`) `test` created automatically, on linux (`Docker version 28.5.0, build 887030f`) it's not.

**Comment by ndeloof** [comment]
@dferenc-netwizzy top-level volume definition is equivalent to a `docker volume create ...` command. Using local driver with ò: bind` you basically run a `mount` command on docker host with your own parameters. I won't recommend using this but for advanced usages.

service-level volume is the same as `docker run --volume` (or --mount, depending selected options). Doing so, engine logic manages the bind mount, including some corner cases. Maybe this windows vs Linux behavior is one of them.
